readtags: optimize the way to read fields

masatake opened the pull request:
I measured performance of readtags by giving large tags file and found
strcmp in parseExtensionFields is a bottle neck.

This change reduces the number of calling strcmp in the function.

Without this change:
[yamato@slave]~/var/ctags-github% time readtags -t ~/.spelunker.d/kernel76/kernel76.tags  -l   > /dev/null (5 times)
<gs -t ~/.spelunker.d/kernel76/kernel76.tags  -l   > /dev/null
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  3.62s user 0.29s system 99% cpu 3.918 total
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  3.62s user 0.30s system 99% cpu 3.924 total
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  3.64s user 0.28s system 99% cpu 3.930 total
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  3.60s user 0.27s system 99% cpu 3.874 total
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  3.65s user 0.25s system 99% cpu 3.909 total

With this change:
[yamato@slave]~/var/ctags-github% time readtags -t ~/.spelunker.d/kernel76/kernel76.tags  -l   > /dev/null (5 times)
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  2.29s user 0.27s system 99% cpu 2.574 total
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  2.24s user 0.27s system 99% cpu 2.518 total
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  2.34s user 0.26s system 99% cpu 2.605 total
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  2.27s user 0.26s system 99% cpu 2.530 total
readtags -t ~/.spelunker.d/kernel76/kernel76.tags -l > /dev/null  2.26s user 0.27s system 99% cpu 2.537 total

ctags with the change is about 1/3 faster than before.

Signed-off-by: Masatake YAMATO <yamato@redhat.com>

b4n commented on read/readtags.c:
```diff
@@ -355,20 +355,38 @@ static void parseExtensionFields (tagFile *const file, tagEntry *const entry,
+						entry->kind = colon + 1;
+					else
+						goto minorField;
+				}
```
I think you need to ask yourself whether the extra performance is worth this kind of code, that'll be a nightmare to maintain (IMO; I can't imagine the pain of adding a new case).

I'd rather go halfway myself, keep the initial computation for the length, but instead of convoluted checking for the key simply use `if (key[0] == 'f' && key[1] == 'i' && key[2] == 'l' && key[3] == 'e')` which will likely be faster than strcmp, maybe even very close to what you have right now, but less complicated.  If that works, you could even make it more readable with a small macro e.g. like `#define CMP4(a, b) ((a)[0] == (b)[0] && (a)[1] == (b)[1] && (a)[2] == (b)[2] && (a)[3] == (b)[3])` and use `if (CMP4(key, "file"))`.

I'd also keep the `value` variable that is used in all paths, unless it's really something the compiler chokes on -- but that's be crazy of it. Make it a constant if you'd like though.

Anyway, all this is not checked for speed, but IMO not all speed gains are worth it if the code becomes too problematic.

b4n replied:
This said, I'd be kind of surprised that a good compiler with optimizations enabled wouldn't do something similar to my `CMP4()` all by itself, and could do that better using specific instruction sets…
Have you tried simply guarding with the length check and keeping the `strcmp()`s?

masatake replied:
> I'd rather go halfway myself, keep the initial computation for the length

I see. My understanding is that you are talking about:
```
if (colon - key == 4
```
.

b4n replied:
Yes.

But remember that I'm just looking at the code, I didn't profile anything, so don't take my word as gospel :wink:

masatake replied:
All you wrote are correct as new result of profiling shows.

masatake commented:
> I'd rather go halfway myself, keep the initial computation for the length

I see. My understanding is that you are talking about:
```
if (colon - key == 4
```
.

masatake commented:
New change using CMP4 is a bit faster than my original change, and CMP4 is maintainable!

```
[yamato@slave]~/var/ctags-github% git checkout master
Already on 'master'
[yamato@slave]~/var/ctags-github% make -j 9
REPOINFO   main/repoinfo.h
make  all-recursive
make[1]: Entering directory '/home/yamato/var/ctags-github'
make[2]: Entering directory '/home/yamato/var/ctags-github'
REPOINFO   main/repoinfo.h
  CC       read/readtags-readtags.o
  CCLD     ctags
  CCLD     readtags
make[2]: Leaving directory '/home/yamato/var/ctags-github'
make[1]: Leaving directory '/home/yamato/var/ctags-github'
[yamato@slave]~/var/ctags-github% cat $t > /dev/null; repeat 5; do time ./readtags -t $t -l > /dev/null; done
./readtags -t $t -l > /dev/null  3.60s user 0.28s system 99% cpu 3.888 total
./readtags -t $t -l > /dev/null  3.65s user 0.26s system 99% cpu 3.932 total
./readtags -t $t -l > /dev/null  3.64s user 0.26s system 99% cpu 3.907 total
./readtags -t $t -l > /dev/null  3.62s user 0.30s system 99% cpu 3.925 total
./readtags -t $t -l > /dev/null  3.62s user 0.26s system 99% cpu 3.888 total
[yamato@slave]~/var/ctags-github% git checkout optimize-readtags-too-much
Switched to branch 'optimize-readtags-too-much'
[yamato@slave]~/var/ctags-github% make -j 9
REPOINFO   main/repoinfo.h
make  all-recursive
make[1]: Entering directory '/home/yamato/var/ctags-github'
make[2]: Entering directory '/home/yamato/var/ctags-github'
REPOINFO   main/repoinfo.h
  CC       read/readtags-readtags.o
  CC       main/ctags-repoinfo.o
  CCLD     ctags
  CCLD     readtags
make[2]: Leaving directory '/home/yamato/var/ctags-github'
make[1]: Leaving directory '/home/yamato/var/ctags-github'
[yamato@slave]~/var/ctags-github% cat $t > /dev/null; repeat 5; do time ./readtags -t $t -l > /dev/null; done
./readtags -t $t -l > /dev/null  2.25s user 0.26s system 99% cpu 2.516 total
./readtags -t $t -l > /dev/null  2.28s user 0.26s system 99% cpu 2.543 total
./readtags -t $t -l > /dev/null  2.27s user 0.27s system 99% cpu 2.542 total
./readtags -t $t -l > /dev/null  2.24s user 0.28s system 99% cpu 2.531 total
./readtags -t $t -l > /dev/null  2.25s user 0.28s system 99% cpu 2.537 total
[yamato@slave]~/var/ctags-github% git checkout optimize-readtags
Switched to branch 'optimize-readtags'
[yamato@slave]~/var/ctags-github% make -j 9
REPOINFO   main/repoinfo.h
make  all-recursive
make[1]: Entering directory '/home/yamato/var/ctags-github'
make[2]: Entering directory '/home/yamato/var/ctags-github'
REPOINFO   main/repoinfo.h
  CC       read/readtags-readtags.o
  CC       main/ctags-repoinfo.o
  CCLD     ctags
  CCLD     readtags
make[2]: Leaving directory '/home/yamato/var/ctags-github'
make[1]: Leaving directory '/home/yamato/var/ctags-github'
[yamato@slave]~/var/ctags-github% cat $t > /dev/null; repeat 5; do time ./readtags -t $t -l > /dev/null; done
./readtags -t $t -l > /dev/null  2.64s user 0.26s system 98% cpu 2.954 total
./readtags -t $t -l > /dev/null  2.68s user 0.26s system 99% cpu 2.948 total
./readtags -t $t -l > /dev/null  2.63s user 0.28s system 99% cpu 2.911 total
./readtags -t $t -l > /dev/null  2.66s user 0.28s system 99% cpu 2.947 total
./readtags -t $t -l > /dev/null  2.69s user 0.26s system 99% cpu 2.960 total
[yamato@slave]~/var/ctags-github% git show |cat
commit 332e0a5fc472b96af07738e10779cf6a4cf81061
Author: Masatake YAMATO <yamato@redhat.com>
Date:   Tue Feb 26 14:47:11 2019 +0900

    readtags: optimize the way to read fields

    I measured performance of readtags by giving large tags file and found
    strcmp in parseExtensionFields is a bottle neck.

    This change reduces the number of calling strcmp in the function.

    Without this change:

    With this change:

    ctags with the change is about 1/3 faster than before.

    Signed-off-by: Masatake YAMATO <yamato@redhat.com>

diff --git a/read/readtags.c b/read/readtags.c
index ebcb5559..22a6d3dd 100644
--- a/read/readtags.c
+++ b/read/readtags.c
@@ -356,15 +356,22 @@ static void parseExtensionFields (tagFile *const file, tagEntry *const entry,
                        {
                                const char *key = field;
                                const char *value = colon + 1;
+                               const int key_len = colon - key;
                                *colon = '\0';
-                               if (strcmp (key, "kind") == 0)
-                                       entry->kind = value;
-                               else if (strcmp (key, "file") == 0)
-                                       entry->fileScope = 1;
-                               else if (strcmp (key, "line") == 0)
-                                       entry->address.lineNumber = atol (value);
+                               if (key_len == 4)
+                               {
+                                       if (strcmp (key, "kind") == 0)
+                                               entry->kind = value;
+                                       else if (strcmp (key, "file") == 0)
+                                               entry->fileScope = 1;
+                                       else if (strcmp (key, "line") == 0)
+                                               entry->address.lineNumber = atol (value);
+                                       else
+                                               goto normalField;
+                               }
                                else
                                {
+                               normalField:
                                        if (entry->fields.count == file->fields.max)
                                                growFields (file);
                                        file->fields.list [entry->fields.count].key = key;
[yamato@slave]~/var/ctags-github% cat $t > /dev/null; repeat 5; do time ./readtags -t $t -l > /dev/null; done
./readtags -t $t -l > /dev/null  2.19s user 0.31s system 99% cpu 2.510 total
./readtags -t $t -l > /dev/null  2.22s user 0.24s system 99% cpu 2.469 total
./readtags -t $t -l > /dev/null  2.23s user 0.26s system 99% cpu 2.495 total
./readtags -t $t -l > /dev/null  2.22s user 0.28s system 99% cpu 2.495 total
./readtags -t $t -l > /dev/null  2.21s user 0.26s system 99% cpu 2.478 total
[yamato@slave]~/var/ctags-github% git diff
diff --git a/read/readtags.c b/read/readtags.c
index 22a6d3dd..e3d0ac77 100644
--- a/read/readtags.c
+++ b/read/readtags.c
@@ -334,6 +334,8 @@ static tagResult growFields (tagFile *const file)
        return result;
 }
 
+#define CMP4(a, b) ((a)[0] == (b)[0] && (a)[1] == (b)[1] && (a)[2] == (b)[2] && (a)[3] == (b)[3])
+
 static void parseExtensionFields (tagFile *const file, tagEntry *const entry,
                                                                  char *const string)
 {
@@ -360,11 +362,11 @@ static void parseExtensionFields (tagFile *const file, tagEntry *const entry,
                                *colon = '\0';
                                if (key_len == 4)
                                {
-                                       if (strcmp (key, "kind") == 0)
+                                       if (CMP4 (key, "kind") == 0)
                                                entry->kind = value;
-                                       else if (strcmp (key, "file") == 0)
+                                       else if (CMP4 (key, "file") == 0)
                                                entry->fileScope = 1;
-                                       else if (strcmp (key, "line") == 0)
+                                       else if (CMP4 (key, "line") == 0)
                                                entry->address.lineNumber = atol (value);
                                        else
                                                goto normalField;
```

masatake commented:
I have to inspect the disassembled code to know how strcmp is optimized.
Anyway I'm happy with CMP4.

masatake commented:
I wonder how many people use readtags.
http://ctags.sourceforge.net/tool_support.html

readtags is yet another great work of Exuberant-ctags.
By adding sorter, filter, and formatter, it can be an incredible tool.

craigbarnes commented:
`memcmp()` should _always_ be used instead of `strcmp()` if you know the length of the string. If `memcmp()` was used here, GCC would use `__builtin_memcmp()`, which would most likely compile down to something even more optimized than the `CMP4` macro.

masatake commented:
@craigbarnes, thank you! Just one cmpl may be much faster.
I will update my change.

masatake commented:
I expected readtags faster with memcmp. However, as far as my measurement, I could not find any more gain. However, using memcmp allows us not to introduce the macro.

masatake commented:
I will merge the commits without squashing because this pull request is so impressive to me. Someone joining to the development of u-ctags in the future will learn the way to optimize the code from these commits.

@b4n and @craigbarnes, thank you very much.
